eppo_client/assignment_logger.py

```diff
@@ -1,13 +1,59 @@
-from typing import Dict
-from eppo_client.base_model import BaseModel
-from pydantic import ConfigDict
+from typing import Dict, Optional, Tuple, MutableMapping
 
 
-class AssignmentLogger(BaseModel):
-    model_config = ConfigDict(arbitrary_types_allowed=True)
-
+class AssignmentLogger:
     def log_assignment(self, assignment_event: Dict):
         pass
 
     def log_bandit_action(self, bandit_event: Dict):
         pass
+
+
+class AssignmentCacheLogger(AssignmentLogger):
+    def __init__(
+        self,
+        inner: AssignmentLogger,
+        *,
+        assignment_cache: Optional[MutableMapping] = None,
+        bandit_cache: Optional[MutableMapping] = None,
+    ):
+        self.__inner = inner
+        self.__assignment_cache = assignment_cache
+        self.__bandit_cache = bandit_cache
+
+    def log_assignment(self, event: Dict):
+        _cache_or_call(
+            self.__assignment_cache,
+            *AssignmentCacheLogger.__assignment_cache_keyvalue(event),
+            lambda: self.__inner.log_assignment(event),
+        )
+
+    def log_bandit_action(self, event: Dict):
+        _cache_or_call(
+            self.__bandit_cache,
+            *AssignmentCacheLogger.__bandit_cache_keyvalue(event),
+            lambda: self.__inner.log_bandit_action(event),
+        )
+
+    @staticmethod
+    def __assignment_cache_keyvalue(event: Dict) -> Tuple[Tuple, Tuple]:
+        key = (event["featureFlag"], event["subject"])
+        value = (event["allocation"], event["variation"])
+        return key, value
+
+    @staticmethod
+    def __bandit_cache_keyvalue(event: Dict) -> Tuple[Tuple, Tuple]:
+        key = (event["flagKey"], event["subject"])
```

Comment:
Noticed that our event format is somewhat inconsistent between SDKs.

In Python, Go, PHP, .NET, Rust, and Ruby:
- assignment event user `"featureFlag"`
- bandit event uses `"flagKey"`

Javascript and Java SDK use `"featureFlag"` for both events.

Reply:
Yeah, it's just a factor of evolution. We're trying to move (slowly) more towards explicit keys, such as `flagKey`. However, it's also nice to have fields be analogously named between the two events (hence JavaScript and Java maintaining `featureFlag`). I think you're just seeing the personal preferences of the authors manifest in the various SDKs.